components/post_view/embedded_bindings/button_binding/button_binding.tsx

```diff
@@ -0,0 +1,127 @@
+        getChannel: (channelId: string) => Promise<ActionResult>;
+    };
+    sendEphemeralPost: (message: string, channelID?: string, rootID?: string) => void;
+}
+
+type State = {
+    executing: boolean;
+}
+
+export class ButtonBinding extends React.PureComponent<Props, State> {
+    constructor(props: Props) {
+        super(props);
+        this.state = {
+            executing: false,
+        };
+    }
+
+    handleClick = async () => {
+        const {binding, post} = this.props;
+        if (!binding.call) {
+            return;
+        }
+
+        let teamID = '';
+        const {data} = await this.props.actions.getChannel(post.channel_id) as {data?: any; error?: any};
+        if (data) {
+            const channel = data as Channel;
+            teamID = channel.team_id;
+        }
+
+        const context = createCallContext(
+            binding.app_id,
+            AppBindingLocations.IN_POST + '/' + binding.location,
+            post.channel_id,
+            teamID,
+            post.id,
+            post.root_id,
+        );
+        const call = createCallRequest(
+            binding.call,
+            context,
+            {post: AppExpandLevels.EXPAND_ALL},
+        );
+        this.setState({executing: true});
+        const res = await this.props.actions.doAppCall(call, AppCallTypes.SUBMIT, this.props.intl);
+
+        this.setState({executing: false});
+        const callResp = (res as {data: AppCallResponse}).data;
+        const ephemeral = (message: string) => this.props.sendEphemeralPost(message, this.props.post.channel_id, this.props.post.root_id);
+        switch (callResp.type) {
+        case AppCallResponseTypes.OK:
+            if (callResp.markdown) {
+                ephemeral(callResp.markdown);
+            }
+            break;
+        case AppCallResponseTypes.ERROR: {
+            const errorMessage = callResp.error || this.props.intl.formatMessage({id: 'apps.error.unknown', defaultMessage: 'Unknown error happenned'});
+            ephemeral(errorMessage);
+            break;
+        }
+        case AppCallResponseTypes.NAVIGATE:
+        case AppCallResponseTypes.FORM:
+            break;
+        default: {
+            const errorMessage = this.props.intl.formatMessage(
+                {id: 'apps.error.responses.unknown_type', defaultMessage: 'App response type not supported. Response type: {type}.'},
+                {type: callResp.type},
+            );
+            ephemeral(errorMessage);
+        }
+        }
+    }
+
+    render() {
+        const {binding} = this.props;
+        let customButtonStyle;
+
+        if (!binding.call) {
+            return null;
+        }
+
+        return (
+            <button
+                onClick={this.handleClick}
+                style={customButtonStyle}
+            >
+                <LoadingWrapper
+                    loading={this.state.executing}
+                >
+                    <Markdown
```

Comment:
@larkox Do you know the reasoning behind using the `Markdown` component here?

Reply:
This is directly taken from https://github.com/mattermost/mattermost-webapp/blob/99e04158c01d5741d19b7dc228751af9972a6825/components/post_view/message_attachments/action_button/action_button.tsx#L68

It was added 2 years ago to support emojis in the buttons: https://github.com/mattermost/mattermost-webapp/commit/1c1ef24740607cac8b41c72829c8c0b47fda9820

Reply:
Ah, back from the period where we were randomly adding Markdown support everywhere 😛. I'm fine leaving it for compatibility then as long as we're aware it could let people make the app look messed up if they really wanted to.